src/rviz/default_plugin/map_display.cpp

```diff
@@ -59,9 +59,162 @@
+  {
+    rend->setCustomParameter( ALPHA_PARAMETER, alpha_vec_ );
+  }
+private:
+  Ogre::Vector4 alpha_vec_;
+};
+
+
+Swatch::Swatch(MapDisplay* parent, unsigned int x, unsigned int y, unsigned int width, unsigned int height, float resolution)
+  : parent_(parent), manual_object_( NULL ), x_(x), y_(y), width_(width), height_(height)
+{
+  // Set up map material
+  static int material_count = 0;
+  std::stringstream ss;
+  ss << "MapMaterial" << material_count++;
+  material_ = Ogre::MaterialManager::getSingleton().getByName("rviz/Indexed8BitImage");
+  material_ = material_->clone( ss.str() );
+
+  material_->setReceiveShadows(false);
+  material_->getTechnique(0)->setLightingEnabled(false);
+  material_->setDepthBias( -16.0f, 0.0f );
+  material_->setCullingMode( Ogre::CULL_NONE );
+  material_->setDepthWriteEnabled(false);
+
+  static int map_count = 0;
+  std::stringstream ss2;
+  ss2 << "MapObject" << map_count++;
+  manual_object_ = parent_->scene_manager_->createManualObject( ss2.str() );
+
+  static int node_count = 0;
+  std::stringstream ss3;
+  ss3 << "NodeObject" << node_count++;
+
+  scene_node_ = parent_->scene_node_->createChildSceneNode(ss3.str());
+  scene_node_->attachObject( manual_object_ );
+
+  manual_object_->begin(material_->getName(), Ogre::RenderOperation::OT_TRIANGLE_LIST);
+  {
+    // First triangle
+    {
+      // Bottom left
+      manual_object_->position( 0.0f, 0.0f, 0.0f );
+      manual_object_->textureCoord(0.0f, 0.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Top right
+      manual_object_->position( 1.0f, 1.0f, 0.0f );
+      manual_object_->textureCoord(1.0f, 1.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Top left
+      manual_object_->position( 0.0f, 1.0f, 0.0f );
+      manual_object_->textureCoord(0.0f, 1.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+    }
+
+    // Second triangle
+    {
+      // Bottom left
+      manual_object_->position( 0.0f, 0.0f, 0.0f );
+      manual_object_->textureCoord(0.0f, 0.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Bottom right
+      manual_object_->position( 1.0f, 0.0f, 0.0f );
+      manual_object_->textureCoord(1.0f, 0.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Top right
+      manual_object_->position( 1.0f, 1.0f, 0.0f );
+      manual_object_->textureCoord(1.0f, 1.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+    }
+  }
+  manual_object_->end();
+
+  scene_node_->setPosition( x*resolution, y*resolution, 0 );
+  scene_node_->setScale( width*resolution, height*resolution, 1.0 );
+
+  if( parent_->draw_under_property_->getValue().toBool() )
+  {
+    manual_object_->setRenderQueueGroup(Ogre::RENDER_QUEUE_4);
+  }
+
+  // don't show map until the plugin is actually enabled
+  manual_object_->setVisible( false );
+
+
+}
+
+Swatch::~Swatch()
+{
+    delete manual_object_;
```

Comment:
I think this should probably use http://www.ogre3d.org/docs/api/1.8/class_ogre_1_1_scene_manager.html#a75d15c3ac882f78f0001d9d007f403be